ts/src/hashkey.ts

```diff
@@ -0,0 +1,4336 @@
+        const marketId = this.safeString (trade, 'symbol');
+        market = this.safeMarket (marketId, market);
+        let side = this.safeStringLower (trade, 'side'); // swap trades have side param
+        if (side !== undefined) {
+            side = this.safeString (side.split ('_'), 0);
+        }
+        const isBuyer = this.safeBool (trade, 'isBuyer');
+        if (isBuyer !== undefined) {
+            side = isBuyer ? 'buy' : 'sell';
+        }
+        let takerOrMaker = undefined;
+        const isMaker = this.safeBool2 (trade, 'isMaker', 'isMarker');
+        if (isMaker !== undefined) {
+            takerOrMaker = isMaker ? 'maker' : 'taker';
+        }
+        let feeCost = this.safeString (trade, 'commission');
+        let feeCurrncyId = this.safeString (trade, 'commissionAsset');
+        const feeInfo = this.safeDict (trade, 'fee');
+        if (feeInfo !== undefined) {
+            feeCost = this.safeString (feeInfo, 'fee');
+            feeCurrncyId = this.safeString (feeInfo, 'feeCoinId');
+        }
+        const fee = {
+            'cost': feeCost,
+            'currency': this.safeCurrencyCode (feeCurrncyId),
+        };
+        return this.safeTrade ({
+            'id': this.safeString2 (trade, 'id', 'tradeId'),
+            'timestamp': timestamp,
+            'datetime': this.iso8601 (timestamp),
+            'symbol': market['symbol'],
+            'side': side,
+            'price': this.safeString2 (trade, 'p', 'price'),
+            'amount': this.safeStringN (trade, [ 'q', 'qty', 'quantity' ]),
+            'cost': undefined,
+            'takerOrMaker': takerOrMaker,
+            'type': undefined,
+            'order': this.safeString (trade, 'orderId'),
+            'fee': fee,
+            'info': trade,
+        }, market);
+    }
+
+    async fetchOHLCV (symbol: string, timeframe = '1m', since: Int = undefined, limit: Int = undefined, params = {}): Promise<OHLCV[]> {
+        /**
+         * @method
+         * @name hashkey#fetchOHLCV
+         * @see https://hashkeyglobal-apidoc.readme.io/reference/get-kline
+         * @description fetches historical candlestick data containing the open, high, low, and close price, and the volume of a market
+         * @param {string} symbol unified symbol of the market to fetch OHLCV data for
+         * @param {string} timeframe the length of time each candle represents
+         * @param {int} [since] timestamp in ms of the earliest candle to fetch
+         * @param {int} [limit] the maximum amount of candles to fetch (max 1000)
+         * @param {object} [params] extra parameters specific to the exchange API endpoint
+         * @param {int} [params.until] timestamp in ms of the latest candle to fetch
+         * @returns {int[][]} A list of candles ordered as timestamp, open, high, low, close, volume
+         */
+        const methodName = 'fetchOHLCV';
+        await this.loadMarkets ();
+        const market = this.market (symbol);
```

Comment:
can we add automatic pagination here? check for instance how bybit does it (it's easy)

Reply:
Done